File: Python/partition-to-k-equal-sum-subsets.py


# Time:  O(k^(n-k) * k!)
# Space: O(n)
# DFS solution with pruning.
class Solution2(object):
    def canPartitionKSubsets(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        def dfs(nums, target, i, subset_sums):
            if i == len(nums):
                return True
            for k in xrange(len(subset_sums)):
                if subset_sums[k]+nums[i] > target:
                    continue
                subset_sums[k] += nums[i]
                if dfs(nums, target, i+1, subset_sums):
                    return True
                subset_sums[k] -= nums[i]
                if not subset_sums[k]: break
            return False

        total = sum(nums)
        if total%k != 0 or max(nums) > total//k:
            return False
        nums.sort(reverse=True)
        subset_sums = [0] * k
        return dfs(nums, total//k, 0, subset_sums)

File: C++/partition-to-k-equal-sum-subsets.cpp
class Solution {
    bool value = false;
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        int sum = 0;
        for(int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }
        
        if (sum % k != 0) {
            return false;
        }
        
        sort(nums.begin(), nums.end(), greater<int>());
        
        vector<int> val (k, 0);
        helper(val, nums, 0, sum / k);
        return value;
    }
    
    void helper(vector<int>&val, vector<int>& nums, int idx, int bucketSize) {
        if (idx == nums.size()) {
            value = true;
            return;
        }
        
        for(int i = 0; i < val.size(); i++) {
            if (val[i] + nums[idx] > bucketSize || value == true) {
                continue;
            }
            
            val[i] += nums[idx];
            helper(val, nums, idx + 1, bucketSize);
            val[i] -= nums[idx];
            
            if (val[i] == 0) break;
        }
    }
};